### PR TITLE
feat(history): add offset pagination for history and statistics sources

### PR DESCRIPTION
**Pull request by Patch76**
Closes #930

## What does this PR do?

Adds `offset` parameter to `ha_get_history` for both sources, and enables `limit` for the `statistics` source (previously only applied to `history`). Both use the standardized `build_pagination_metadata` helper already established in `tools_search`, `tools_services`, `tools_integrations`, and `tools_registry`.

`ha_get_history(source="history")` had `limit` but no `offset` — impossible to access older state changes once the limit was exceeded. `ha_get_history(source="statistics")` had neither — 30 days of 5-minute intervals yields 8,640 rows per entity, which could exhaust LLM context.

Both sources now return `total_count`, `offset`, `limit`, `count`, `has_more`, `next_offset`. HA WebSocket APIs have no native server-side offset/limit — all data is fetched before slicing. `limit` and `offset` apply per entity, not globally across entities.

Renamed response fields for consistency with `build_pagination_metadata`: `total_available` → `total_count`, `truncated` → `has_more` (history source); `count` promoted to pagination metadata (statistics source).

## Type of change
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 📚 Documentation
- [ ] 🔧 Maintenance/refactor
- [ ] 🧪 Tests only
- [ ] 💥 Breaking change

## Testing
- [ ] I have tested these changes with a LLM agent
- [x] All automated tests pass (`uv run pytest`)
- [x] Code follows style guidelines (`uv run ruff check`)

## Checklist
- [ ] I have updated documentation if needed

### CODE REVIEW

File: src/ha_mcp/tools/tools_history.py [hunk 1]
```diff
@@ -391,6 +408,23 @@ async def _fetch_history(
...
 
+    try:
+        effective_offset = coerce_int_param(
+            offset,
+            param_name="offset",
+            default=0,
+            min_value=0,
+        )
+        if effective_offset is None:
+            effective_offset = 0
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_history.py]: Fixed in 3a0a2b8. Removed the redundant `if effective_offset is None` checks in both `_fetch_history` and `_fetch_statistics`. You're correct — `coerce_int_param` with `default=0` (typed as `int` in the `@overload` signature) is guaranteed to return `int`, never `None`.

File: src/ha_mcp/tools/tools_history.py [hunk 2]
```diff
@@ -478,8 +516,44 @@ async def _fetch_statistics(
...
+
+    try:
+        effective_offset = coerce_int_param(
+            offset,
+            param_name="offset",
+            default=0,
+            min_value=0,
+        )
+        if effective_offset is None:
+            effective_offset = 0
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_history.py]: Fixed in 3a0a2b8. Removed the redundant `if effective_offset is None` checks in both `_fetch_history` and `_fetch_statistics`. You're correct — `coerce_int_param` with `default=0` (typed as `int` in the `@overload` signature) is guaranteed to return `int`, never `None`.

File: src/ha_mcp/tools/tools_history.py [hunk 3]
```diff
@@ -478,8 +516,44 @@ async def _fetch_statistics(
...
     """Execute the recorder/statistics_during_period WebSocket call."""
+    # Coerce limit and offset
+    try:
+        effective_limit = coerce_int_param(
+            limit,
+            param_name="limit",
+            default=_DEFAULT_HISTORY_LIMIT,
+            min_value=1,
+            max_value=_MAX_HISTORY_LIMIT,
+        ) or _DEFAULT_HISTORY_LIMIT
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_history.py]: Fixed in 3a0a2b8. Removed the redundant `or _DEFAULT_HISTORY_LIMIT` — `coerce_int_param` with `default=_DEFAULT_HISTORY_LIMIT` already guarantees a non-None return.

File: src/ha_mcp/tools/tools_history.py [hunk 4]
```diff
@@ -391,6 +408,21 @@ async def _fetch_history(
...
+            default=0,
+            min_value=0,
+        )
+    except ValueError as e:
+        raise_tool_error(create_error_response(
+            ErrorCode.VALIDATION_INVALID_PARAMETER,
+            str(e),
+            context={"parameter": "offset"},
+            suggestions=["Provide offset as a non-negative integer (e.g., 0)"],
+        ))
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_history.py:453]: The HA WebSocket API (history/history_during_period) has no native offset parameter — pagination is necessarily client-side. coerce_int_param is required here because AI tools frequently pass numeric parameters as strings (e.g., offset="50"), which would cause a TypeError on the slice operation. This is the established pattern across all paginated tools in the repo (tools_search, tools_services, tools_integrations, tools_registry). There is no backend to delegate to for this validation.

File: src/ha_mcp/tools/tools_history.py [hunk 5]
```diff
@@ -478,8 +514,42 @@ async def _fetch_statistics(
...
+            min_value=1,
+            max_value=_MAX_HISTORY_LIMIT,
+        )
+    except ValueError as e:
+        raise_tool_error(create_error_response(
+            ErrorCode.VALIDATION_INVALID_PARAMETER,
+            str(e),
+            context={"parameter": "limit"},
+            suggestions=["Provide limit as an integer (e.g., 100)"],
+        ))
```
[the remark this answers is not in the document]
Comment (Patch76) [reply src/ha_mcp/tools/tools_history.py:565]: Same reasoning as above — the HA WebSocket backend does not support server-side limit for history/statistics queries. The validation is client-side by necessity, and the coerce_int_param pattern (handling string inputs from AI tools) is established repo-wide. Removing it would break the tool when limit is passed as a string.

**Review by kingpanther13 — CHANGES_REQUESTED** [review]
Thanks for tackling #930 — this is a real gap, especially for statistics where 8,640 rows/entity is easy to hit. A few things to sort out before merge.

## Main concern: per-entity pagination semantics

Every other call site of `build_pagination_metadata` in the repo (`tools_search`, `tools_services`, `tools_integrations`, `tools_registry`) paginates a single flat list at the top level. This PR is the first to apply it per-sublist with no top-level aggregate, so `limit=100` across 5 entities returns up to 500 rows and callers have no top-level `has_more` to drive a single pagination loop. For token-budget use cases (the main motivation for the shared helper), that's a footgun.

Either approach would work:
1. **Reject `offset > 0` when `len(entity_id_list) > 1`** with `VALIDATION_INVALID_PARAMETER` ("offset pagination requires a single entity_id"). Simplest, keeps the response shape aligned with every other pagination site.
2. **Keep per-entity blocks but add a top-level aggregate** (`total_count = sum(...)`, `has_more = any(...)`) alongside them, so a single-loop caller has one signal to check.

Option 1 is my preference — it collapses the mental model back to what the other four PRs established.

## Test coverage

`test_history_pagination.py` only exercises single-entity cases, so the PR's headline per-entity behavior is entirely untested. At minimum add a case where entity A has ~50 rows and entity B has ~500, verify `limit=100, offset=0` returns the right per-entity `has_more`/`next_offset` divergence. Per AGENTS.md, E2E is preferred for tool changes — an offset case in `tests/src/e2e/workflows/core/test_history.py` would also catch field-rename drift against real HA payloads. (If you go with option 1 above, the multi-entity test becomes a negative test for the validation error instead.)

Also missing: a default-limit test for the history source (the statistics branch has one at `test_default_limit_applied`), and a `limit` boundary case around `_MAX_HISTORY_LIMIT = 1000`.

## Smaller items

- `tools_history.py:402` — `coerce_int_param(..., default=default_limit, ...) or default_limit` is redundant: `default=default_limit` already handles `None`, and `min_value=1` prevents `0`. `_fetch_statistics` at L523 doesn't have the trailing `or` — inconsistent. Drop it.
- `tools_history.py:186-192` — `offset: int | str` vs `limit: int | str | None`. Align the annotations (either both allow `None` or commit to concrete defaults on both).
- `tools_history.py:499-503` — `query_params` echoes `limit` but not `offset`. Add `"offset": effective_offset` for round-trip observability.
- `tools_history.py:182` — the `limit` `Field` description should note per-entity semantics directly, since schema consumers won't see the docstring NOTE. Something like: "Applied per entity; total rows can reach `limit * len(entity_ids)`."
- `tools_history.py:232-234` — if you keep per-entity semantics (option 2 above), promote `**NOTE:**` to `**WARNING:**` and state the multiplication concretely — per AGENTS.md, non-obvious gotchas warrant WARNING.
- `tools_history.py:521` — `# Coerce limit and offset` just restates the code below it. Drop per "comments only for non-obvious logic."

## What's good

- Shared-helper reuse and field-rename (`total_available`→`total_count`, `truncated`→`has_more`) land cleanly — no stale references anywhere.
- Error handling correctly follows Pattern B with the outer `except ToolError: raise` guard intact.
- Unit tests use a `PAGINATION_FIELDS` set assertion to lock response schema — nice.
- Docstring examples cover both sources with concrete `limit`/`offset` values.

**Comment by Patch76** [comment]
Addressed in d352788 (6 commits). Summary:

## Main concern: per-entity pagination semantics

Went with **Option 1**: `offset > 0` is now rejected with `VALIDATION_INVALID_PARAMETER` when `len(entity_id_list) > 1`. Guard is placed in `ha_get_history` before any WebSocket call, so no data is fetched before the error fires.

I chose `offset: int | str | None = None` (Option B) rather than `int | str = 0` to stay consistent with `limit: int | str | None = None` — both parameters now follow the same nullable pattern and `coerce_int_param` handles the `None → default` coercion in both cases.

## Test coverage

- `TestMultiEntityOffsetGuard` (unit): `offset > 0` rejected, `offset=0` allowed for multi-entity
- `TestHistoryLimitBoundary` (unit): default limit for history source (was missing, statistics had it), `limit > 1000` raises `VALIDATION_INVALID_PARAMETER` — `coerce_int_param(max_value=1000)` raises, no silent cap
- `test_offset_pagination_single_entity` (E2E): two-page walk, verifies `offset`/`total_count` consistency across pages
- `test_multi_entity_offset_rejected` (E2E): guard verified against live HA

## Smaller items

All addressed:
- `or default_limit` removed from `_fetch_history` (redundant after `coerce_int_param`)
- `offset` annotation unified to `int | str | None`
- `offset` added to `query_params` in `_fetch_history`
- `limit` `Field` description notes the multi-entity constraint
- `**NOTE:**` promoted to `**WARNING:**` with explicit multi-entity constraint
- Trivial `# Coerce limit and offset` comment dropped
- `extract_tools.py` run — `offset` parameter now present in `site/src/data/tools.json`, `README.md`, and `DOCS.md`

**Comment by Patch76** [comment]
One follow-up: the referenced SHA in my previous comment was d352788 (before CI feedback). The actual final commit is 43e28a3 — it corrects `test_limit_exceeds_max_raises_error` to `test_limit_exceeds_max_is_clamped`, since `coerce_int_param(max_value=1000)` soft-caps rather than raises. All 15 CI checks are green on that commit.

**Review by kingpanther13 — CHANGES_REQUESTED** [review]
One last thing before merge — the multi-entity offset guard at `tools_history.py` L259 uses raw `int(offset)`:

```python
if offset is not None and offset != 0:
    _effective_offset_check = int(offset) if isinstance(offset, str) else offset
    if _effective_offset_check > 0 and len(entity_id_list) > 1:
        raise_tool_error(...)
```

If an agent passes `offset="garbage"` with multiple entity_ids, `int("garbage")` raises `ValueError` before the multi-entity check runs. That gets caught by the outer `except Exception` and classified into a generic error, bypassing the clean `VALIDATION_INVALID_PARAMETER` response that `coerce_int_param` emits for bad string input.

AI agents hallucinate garbage params more than one might assume, so I'd rather not leave a path where invalid offset produces an inconsistent error shape.

Simplest fix: reuse `coerce_int_param` in the guard itself so both validation and coercion share one code path:

```python
try:
    _effective_offset_check = coerce_int_param(
        offset, param_name="offset", default=0, min_value=0,
    )
except ValueError as e:
    raise_tool_error(create_error_response(
        ErrorCode.VALIDATION_INVALID_PARAMETER,
        str(e),
        context={"parameter": "offset"},
        suggestions=["Provide offset as a non-negative integer"],
    ))

if _effective_offset_check > 0 and len(entity_id_list) > 1:
    raise_tool_error(...)
```

Add a unit test (`test_multi_entity_invalid_string_offset`) asserting `offset="garbage"` with multiple entities raises `VALIDATION_INVALID_PARAMETER`, not a generic error.

Everything else is good to go — approving as soon as this is in.

**Comment by Patch76** [comment]
Fixed in 7b0db7d — replaced the raw `int(offset)` cast with `coerce_int_param` in the multi-entity guard, wrapped in try/except that raises `VALIDATION_INVALID_PARAMETER`. Added `test_multi_entity_invalid_string_offset` asserting the clean error shape for `offset="garbage"` with multiple entity_ids.

Also clarified the `limit` Field description to note that total rows can reach `limit × len(entity_ids)` for multi-entity requests.

**Comment by kingpanther13** [comment]
Btw, this might come in handy when you're doing multiple commits/changes after a review requested changes. Clicking this will let us know you're done and ready for us to look again :)
Screenshot_20260413_070905_Chrome

**Comment by kingpanther13** [comment]
Going to do another Gemini review since there are so many new commits

/Gemini review

File: src/ha_mcp/tools/tools_history.py [hunk 6]
```diff
@@ -221,23 +229,59 @@ async def ha_get_history(
...
+                    default=0,
+                    min_value=0,
+                )
+            except ValueError as e:
+                raise_tool_error(create_error_response(
+                    ErrorCode.VALIDATION_INVALID_PARAMETER,
+                    str(e),
+                    context={"parameter": "offset"},
+                    suggestions=["Provide offset as a non-negative integer (e.g., 0)"],
+                ))
```
[the remark this answers is not in the document]
Comment (kingpanther13) [reply src/ha_mcp/tools/tools_history.py:276]: Fair observation — `offset` is now coerced in two places. It's duplication rather than a correctness issue (both coerce calls use identical semantics), so leaving it for now. If it starts causing drift or maintenance pain later, we'll hoist it then.

**Review by kingpanther13 — APPROVED** [review]
7b0db7d resolves the garbage-offset gap — multi-entity guard now uses `coerce_int_param` and emits a clean `VALIDATION_INVALID_PARAMETER` for invalid string offsets, with a unit test pinning the error shape. Limit-description clarification on multi-entity row totals is a nice touch.
